agents/self_evolving_car/train.py

```diff
@@ -0,0 +1,163 @@
+
+        # RENDER RESULTS
+        action_display = "GEN: "+str(self.gen+1)+" | BOT: "+str(self.brain)#+" \nTURN: "+str(self.botList[self.brain].Nodes[1].node(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,4,5,6,7))+" \nBOOST: "+str(self.botList[self.brain].Nodes[2].node(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,8,9,10,11))+" \nTHROTTLE: "+str(self.botList[self.brain].Nodes(self.distance_to_ball_x,self.distance_to_ball_y,self.distance_to_ball_z,0,1,2,3))
+
+        draw_debug(self.renderer, my_car, packet.game_ball, action_display)
+
+        # # STOP EVOLVING WHEN THE BALL IS TOUCHED
+        # if packet.game_ball.latest_touch.player_name == "Self-Evolving-Car":
+        #     self.mut_rate = 0
+
+        # GAME STATE
+        car_state = CarState(boost_amount=100)
+        ball_state = BallState(Physics(velocity=Vector3(0, 0, 0), location=Vector3(0, -1000, 1200), angular_velocity=Vector3(0, 0, 0)))
+        game_state = GameState(ball=ball_state, cars={self.index: car_state})
+        self.set_game_state(game_state)
+
+        # NEURAL NET OUTPUTS
+        with self.torch.no_grad():
+            outputs = self.bot_list[self.brain].forward(*inputs)
+        self.controller_state = self.output_formatter.format_model_output(outputs, [packet])[0]
+
+        # KILL
+        if (my_car.physics.location.z < 100 or my_car.physics.location.z > 1950 or my_car.physics.location.x < -4000 or my_car.physics.location.x > 4000 or my_car.physics.location.y > 5000) and self.frame > 50:
+            #self.botList[self.brain].fitness = (1/self.frame+1)*100000
+            self.frame = self.max_frames
+
+        # LOOPS
+        self.frame = self.frame + 1
+
+        if self.frame >= self.max_frames:
+            self.frame = 0
+            self.calc_fitness()
+            self.brain += 1  # change bot every reset
+            self.controller_state = SimpleControllerState()  # reset controller
+            self.reset()  # reset at start
+
+        if self.brain >= self.pop:
+            self.gen += 1
+            self.brain = 0  # reset bots after all have gone
+
+            self.avg_best_fitness()
+            self.calc_fittest()
+
+            # PRINT GENERATION INFO
+            print("")
+            print("     GEN = "+str(self.gen))
+            print("-------------------------")
+            print("FITTEST = BOT " + str(self.fittest))
+            print("------FITNESS = " + str(self.bot_fitness[self.fittest]))
+            # print("------WEIGHTS = " + str(self.bot_list[self.fittest]))
+            for i in range(len(self.bot_list)):
+                print("FITNESS OF BOT " + str(i) + " = " + str(self.bot_fitness[i]))
+
+            # NE Functions
+
+            self.selection()
+            self.mutate()
+            self.reset()  # reset because of delay
+
+        return self.controller_state
+
+    def calc_fitness(self):
```

Comment:
We should move this to its own file in framework.
That way we can use fitness calculation in a general way
Along with a lot of the other parts of calculating fitness and mutating a model

That way the agent can just deal with running the model while we mutate it and switch the model it is using in a separate place.
This will work better long term if we ever want more than one genetic agent at the same time controlled by a single central model